File: actions/log_action.py
"""LogAction — parses and prints one-line packet summaries."""

import struct

from actions.base import Action
from core.constants import (
    CLASS_NAMES,
    OFFSET_PROTOCOL,
    OFFSET_SRC_PORT,
    OFFSET_DST_PORT,
)


# IP header: src IP at bytes 26-29, dst IP at bytes 30-33 (standard Ethernet+IP)
_OFFSET_SRC_IP = 26
_OFFSET_DST_IP = 30


def _parse_ip(raw: bytes, offset: int) -> str:
    """Extract dotted-quad IP from raw packet bytes."""
    if len(raw) < offset + 4:
        return "?.?.?.?"
    return ".".join(str(b) for b in raw[offset : offset + 4])


def _parse_port(raw: bytes, offset: int) -> int:
    """Extract big-endian 16-bit port from raw packet bytes."""
    if len(raw) < offset + 2:
        return 0
    return struct.unpack("!H", raw[offset : offset + 2])[0]


class LogAction(Action):
    """Logs one-line packet summaries to stdout."""

    def __init__(self, verbose: bool = False):
        self.verbose = verbose
        self._count = 0

    def execute(self, packet: bytes, class_id: int) -> None:
        self._count += 1
        if not self.verbose and self._count > 10:
            return

        label = CLASS_NAMES.get(class_id, f"CLASS_{class_id}")
        src_ip = _parse_ip(packet, _OFFSET_SRC_IP)
        dst_ip = _parse_ip(packet, _OFFSET_DST_IP)
        src_port = _parse_port(packet, OFFSET_SRC_PORT)
        dst_port = _parse_port(packet, OFFSET_DST_PORT)

        print(
            f"[{label}] {src_ip}:{src_port} -> {dst_ip}:{dst_port} "
            f"({len(packet)} bytes)"
        )

    def close(self) -> None:
        if not self.verbose and self._count > 10:
            print(f"... ({self._count - 10} more packets not shown)")
        print(f"LogAction: {self._count} packets total")

File: graphos-cpp/src/actions/log_action.cpp
#include "graphos/actions/log_action.hpp"
#include "graphos/core/constants.hpp"
#include <cstdio>

namespace graphos {

std::string LogAction::parse_ip(const uint8_t* data, size_t len, int offset) {
    if (static_cast<size_t>(offset + 4) > len)
        return "?.?.?.?";
    char buf[16];
    std::snprintf(buf, sizeof(buf), "%u.%u.%u.%u",
                  data[offset], data[offset+1],
                  data[offset+2], data[offset+3]);
    return buf;
}

uint16_t LogAction::parse_port(const uint8_t* data, size_t len, int offset) {
    if (static_cast<size_t>(offset + 2) > len)
        return 0;
    return static_cast<uint16_t>((data[offset] << 8) | data[offset+1]);
}

void LogAction::execute(const OwnedPacket& packet, int class_id) {
    count_++;
    if (!verbose_ && count_ > 10) return;

    auto it = CLASS_NAMES.find(class_id);
    const char* label = (it != CLASS_NAMES.end()) ? it->second.c_str() : "UNKNOWN";

    auto src_ip = parse_ip(packet.bytes.data(), packet.length, OFFSET_SRC_IP);
    auto dst_ip = parse_ip(packet.bytes.data(), packet.length, OFFSET_DST_IP);
    auto src_port = parse_port(packet.bytes.data(), packet.length, OFFSET_SRC_PORT);
    auto dst_port = parse_port(packet.bytes.data(), packet.length, OFFSET_DST_PORT);

    std::printf("[%s] %s:%u -> %s:%u (%zu bytes)\n",
                label, src_ip.c_str(), src_port,
                dst_ip.c_str(), dst_port, packet.length);
}

void LogAction::close() noexcept {
    if (!verbose_ && count_ > 10) {
        std::printf("... (%zu more packets not shown)\n", count_ - 10);
    }
    std::printf("LogAction: %zu packets total\n", count_);
}

} // namespace graphos
